util/expensivequery/expensivequery.go

```diff
@@ -73,12 +81,208 @@ func (eqh *Handle) Run() {
+		return
+	}
+
+	if oomRecordStatus == uninitialized {
+		if alert := config.GetGlobalConfig().Performance.ServerMemoryAlert; alert == 0 || alert == 1 {
+			oomRecordStatus = 3
+			return
+		}
+		if serverMemoryQuota == 0 {
+			if config.GetGlobalConfig().Performance.ServerMemoryQuota != 0 {
+				serverMemoryQuota = config.GetGlobalConfig().Performance.ServerMemoryQuota
+				oomRecordStatus = useInstanceMemory
+			} else {
+				serverMemoryQuota, oomRecordErr = memory.MemTotal()
+				if oomRecordErr != nil {
+					logutil.BgLogger().Warn("Get system total memory fail.", zap.Error(oomRecordErr))
+					return
+				}
+				oomRecordStatus = useSystemMemory
+			}
+		}
+		lastOOMtime = time.Time{}
+	}
+
+	var memoryUsage uint64
+	instanceStats := &runtime.MemStats{}
+	if oomRecordStatus == 1 {
+		runtime.ReadMemStats(instanceStats)
+		memoryUsage = instanceStats.HeapAlloc
+	} else {
+		memoryUsage, oomRecordErr = memory.MemUsed()
+		if oomRecordErr != nil {
+			logutil.BgLogger().Warn("Get system usage memory fail.", zap.Error(oomRecordErr))
+			return
+		}
+	}
+
+	if float64(memoryUsage) > float64(serverMemoryQuota)*config.GetGlobalConfig().Performance.ServerMemoryAlert ||
+		(oomRecordStatus == useInstanceMemory && instanceStats.NextGC > serverMemoryQuota) {
+		// At least ten seconds between two recordings that memory usage is less than threshold (default 80% system memory).
+		// If the memory is still exceeded, only records once.
+		if time.Since(lastOOMtime) > 10*time.Second {
+			eqh.oomRecord(memoryUsage, serverMemoryQuota, oomRecordStatus == useSystemMemory)
+		}
+		lastOOMtime = time.Now()
+	}
+}
+
+func (eqh *Handle) oomRecord(memUsage uint64, serverMemoryQuota uint64, useSystemMemory bool) {
+	var err error
+	if tmpDir == "" {
+		tmpDir, err = ioutil.TempDir("", "TiDBOOM")
+		if err != nil {
+			return
+		}
+	}
+	tryRemove := func(filename []string) {
+		// Keep the last 5 files
+		if len(filename) < 5 {
+			return
+		}
+		err = os.Remove(filename[0])
+		filename = filename[1:]
+	}
+	tryRemove(lastLogFileName)
+	tryRemove(lastProfileFileName)
+
+	atomic.AddUint32(&OOMRecordCount, 1)
+	if useSystemMemory {
+		logutil.BgLogger().Warn("The os system now takes a lot of memory, has the risk of OOM",
+			zap.Any("osMemUsage", memUsage),
+			zap.Any("osMemoryTotal", serverMemoryQuota),
+		)
+	} else {
+		logutil.BgLogger().Warn("The TiDB instance now takes a lot of memory, has the risk of OOM",
+			zap.Any("tidbMemUsage", memUsage),
+			zap.Any("serverMemoryQuota", serverMemoryQuota),
+		)
+	}
+
+	eqh.oomRecordSQL()
+	eqh.oomRecordProfile()
+}
+
+func (eqh *Handle) oomRecordSQL() {
+	sm := eqh.sm.Load().(util.SessionManager)
+	processInfo := sm.ShowProcessList()
+	pinfo := make([]*util.ProcessInfo, 0, len(processInfo))
+	for _, info := range processInfo {
+		if len(info.Info) != 0 {
+			pinfo = append(pinfo, info)
+		}
+	}
+	now := time.Now()
+
+	fileName := filepath.Join(tmpDir, "oom_sql"+time.Now().Format(time.RFC3339))
```

Comment:
how about add pid to this filename and also print pid in "Welcome to TiDB" log.

maybe the user will deploy multi tidb in one machine, we need identity which log belongs to which process

Reply:
I use `ioutil.TempDir` to create `tmpDir`, and it will use different random suffix for each instance. So the `fileName` can determine which log file generated by the instance and it is printed in `tidb.log`. Is the pid necessary?